xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenOrderBookUtils.java

```diff
@@ -0,0 +1,100 @@
+            Map<String, List<List<String>>> orderBookItemsMap = new HashMap(orderBookItems);
+            int index = 2;
+            if (jsonParseResult.size() > 4) {
+                orderBookItems = getTypedValue(iterator, Map.class, "order book items");
+                orderBookItemsMap.putAll(orderBookItems);
+                index = 3;
+            }
+            String channelName = (String) jsonParseResult.get(index++);
+            String pair = (String) jsonParseResult.get(index);
+            KrakenOrderBookMessageType orderBookType = orderBookItemsMap.keySet().stream()
+                    .anyMatch(key -> StringUtils.equalsAny(key, ASK_SNAPSHOT, BID_SNAPSHOT)) ? KrakenOrderBookMessageType.SNAPSHOT : KrakenOrderBookMessageType.UPDATE;
+
+            List<List<String>> asksValues;
+            List<List<String>> bidsValues;
+            if (orderBookType == KrakenOrderBookMessageType.SNAPSHOT) {
+                asksValues = orderBookItems.get(ASK_SNAPSHOT);
+                bidsValues = orderBookItems.get(BID_SNAPSHOT);
+            } else {
+                asksValues = orderBookItems.get(ASK_UPDATE);
+                bidsValues = orderBookItems.get(BID_UPDATE);
+            }
+            return new KrakenOrderBook(channelID, channelName, pair, orderBookType, getItemsArray(asksValues), getItemsArray(bidsValues));
+
+        } catch (KrakenException e) {
+            LOG.error("failed to parse order book tree {}", e.getMessage());
+            return null;
+        }
+    }
+
+    public static KrakenPublicOrder[] getItemsArray(List<List<String>> values) {
+        return values == null ? new KrakenPublicOrder[0] : values.stream().map(KrakenOrderBookUtils::extractKrakenPublicOrder).toArray(KrakenPublicOrder[]::new);
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T> T getTypedValue(Iterator iterator, Class<T> clazz, String fieldName) throws KrakenException {
+        if (!iterator.hasNext()) {
+            throw new KrakenException(String.format("Expected value of %s type for %s filed but there is no value", clazz, fieldName));
+        }
+        Object object = iterator.next();
+        if (!clazz.isAssignableFrom(object.getClass())) {
+            throw new KrakenException(String.format("Expected value of %s type for %s filed but there is invalid type %s", clazz, fieldName, object.getClass().getName()));
+        }
+        return (T) object;
+    }
+
+    public static KrakenPublicOrder extractKrakenPublicOrder(List<String> list) {
+        return new KrakenPublicOrder(
+                new BigDecimal(list.get(0)).stripTrailingZeros(),
+                new BigDecimal(list.get(1)).stripTrailingZeros(),
+                timestampToMs(list.get(2))
+        );
+    }
+
+    private static long timestampToMs(String timestamp) {
+        return new BigDecimal(timestamp).multiply(new BigDecimal(1000).abs()).longValue();
```

Comment:
In this place abs() invers only new BigDecimal(1000)

Reply:
I have removed abs - its not required here at all.